File: js/src/core/notification.ts
import { NotificationMessage, NotificationPayload } from '../types';
import { MissingRequiredParameterError } from '../errors';

/**
 * Value object describing a notification to send to a specific user.
 *
 * Carries the target userId plus the message (title, body, optional link) and
 * serializes via `toJSON()` to the payload the platform signs and sends.
 * Constructed by callers; consumed by YeriaSigner.signNotification.
 */
export class Notification {
    private userId: string;
    private message: NotificationMessage;

    constructor(userId: string, title: string, body: string, link?: string) {
        if (!userId || !title || !body) {
            throw new MissingRequiredParameterError('userId, title, and body');
        }
        this.userId = userId;
        this.message = { title, body, link };
    }

    /**
     * Sets the optional in-app navigation link
     * @param link - Link URL (e.g., "/profile" or "app://view/123")
     * @returns this for chaining
     */
    setLink(link: string): this {
        this.message.link = link;
        return this;
    }

    /**
     * Returns the notification payload as JSON
     */
    toJSON(): NotificationPayload {
        return {
            userId: this.userId,
            message: this.message
        };
    }
}

File: py/yeriasdk/core/notification.py
"""
Notification class for sending user notifications
"""

from typing import Optional
from ..types.models import NotificationMessage, NotificationPayload
from ..errors.exceptions import MissingRequiredParameterError


class Notification:
    """Value object describing a notification to send to a specific user.

    Carries the target user_id plus the message (title, body, optional link)
    and serializes via ``to_json()`` to the payload the platform signs and
    sends. Constructed by callers; consumed by YeriaSigner.sign_notification.
    """

    def __init__(self, user_id: str, title: str, body: str, link: Optional[str] = None):
        if not user_id or not title or not body:
            raise MissingRequiredParameterError("userId, title, and body")

        self.user_id = user_id
        self.message = NotificationMessage(title=title, body=body, link=link)

    def set_link(self, link: str) -> "Notification":
        """Set the optional in-app navigation link"""
        self.message.link = link
        return self

    def to_json(self) -> NotificationPayload:
        """Returns the notification payload as JSON"""
        return NotificationPayload(user_id=self.user_id, message=self.message)


